lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,378 @@
+          interfaceMap[key] = {Component::Type::Bundle, description};
+        else if (tpe == opts.augmentedVectorType)
+          interfaceMap[key] = {Component::Type::Vector, description};
+        else
+          interfaceMap[key] = {Component::Type::Ground, description};
+        interfaceKeys.push_back(key);
+      }
+      // Create the interface.
+      if (interfaces.count(defName.getValue()) == 0) {
+        auto interfaceOp =
+            builder.create<sv::InterfaceOp>(circuitOp->getLoc(), defName);
+        ImplicitLocOpBuilder blockBuilder(interfaceOp->getLoc(),
+                                          interfaceOp.getRegion());
+        blockBuilder.createBlock(&interfaceOp.getRegion());
+        blockBuilder.create<sv::TypeDeclTerminatorOp>();
+        interfaces[defName.getValue()] = interfaceOp;
+      }
+      continue;
+    }
+
+    unprocessedAnnos.push_back(a);
+  }
+
+  if (unprocessedAnnos.empty())
+    circuitOp->removeAttr("annotations");
+  else
+    circuitOp->setAttr("annotations", builder.getArrayAttr(unprocessedAnnos));
+
+  for (auto &op : circuitOp.getBody()->getOperations()) {
+    TypeSwitch<Operation *>(&op).Case<FModuleOp, FExtModuleOp>([&](auto op) {
+      return GrandCentralVisitor(&opts, &getContext(), &interfaceMap)
+          .visitDecl(op);
+    });
+  }
+
+  auto loc = UnknownLoc::get(context);
+
+  // Populate the interfaces, iterating in reverse order due as statements are
+  // inserted at the front of the Interface region.
+  for (auto i = interfaceKeys.size(); i != 0; --i) {
+    auto a = interfaceKeys[i - 1];
+    auto defName = a.first;
+    auto name = a.second;
+
+    auto info = &interfaceMap[{defName, name}];
+    if (!info->found && info->tpe != Component::Type::Bundle) {
+      circuitOp.emitOpError()
+          << "contained a Grand Central Interface with defName '" << defName
+          << "' that had a component called '" << name
+          << "' which did not have a scattered companion annotation.  This may "
+             "indicate that the companion was improperly deleted.";
+      continue;
+    }
+
+    auto description = info->description;
+
+    auto interfaceOp = interfaces[defName];
+    builder.setInsertionPoint(&interfaceOp.body().front(),
+                              interfaceOp.body().front().begin());
+
+    if (!description.empty())
+      builder.create<sv::VerbatimOp>(loc, ("\n// " + description).str());
+
+    switch (info->tpe) {
+    case Component::Type::Bundle:
+      // TODO: Change this to actually use an interface type.  This currently
+      // does not work because: (1) interfaces don't have a defined way to get
+      // their bit width and (2) interfaces have a symbol table that is used to
+      // verify internal ops, but this requires looking arbitrarily far upwards
+      // to find other symbols.
+      interfaceOp.emitWarning()
+          << "Nested interfaces are currently not supported! This will be "
+             "jankily verbatim'd out.\n";
```

Comment:
Warnings and error message diagnostics conventionally start with a lower case letter and don't end with a period:
https://clang.llvm.org/docs/InternalsManual.html#the-format-string

Reply:
Cleaned up to use this convention. This thing begs for `emitOpWarning` (which doesn't exist).